librad/src/git/storage.rs

```diff
@@ -136,51 +136,106 @@ impl Storage {
     where
         Oid: AsRef<git2::Oid> + Debug,
     {
-        let oid = oid.as_ref();
-        if oid.is_zero() {
-            // XXX: should this be a panic or error?
-            tracing::warn!("zero oid");
-            return Ok(false);
-        }
+        let oid = match self.has_object(oid)? {
+            Some(oid) => oid,
+            None => return Ok(false),
+        };
 
         self.backend
-            .find_commit(*oid)
+            .find_commit(oid)
             .and_then(|commit| {
-                let namespace = Namespace::from(urn);
-                let branch = {
-                    let path = match &urn.path {
-                        Some(refl) => refl.as_str(),
-                        None => "rad/id",
-                    };
-                    path.strip_prefix("refs/").unwrap_or(path)
+                let tip = match self.get_urn_oid(urn)? {
+                    Some(tip) => tip,
+                    None => return Ok(false),
                 };
-                self.backend
-                    .refname_to_id(&format!("refs/namespaces/{}/refs/{}", namespace, branch))
-                    .and_then(|tip| {
-                        Ok(tip == commit.id()
-                            || self.backend.graph_descendant_of(tip, commit.id())?)
-                    })
-                    .or_matches(is_not_found_err, || Ok(false))
+                Ok(tip == commit.id() || self.backend.graph_descendant_of(tip, commit.id())?)
+            })
+            .or_matches(is_not_found_err, || {
+                tracing::debug!("commit not found");
+                Ok(false)
             })
-            .or_matches(is_not_found_err, || Ok(false))
+    }
+
+    /// Check the existence of `oid` as a **tag**.
+    ///
+    /// The result will be `false` if:
+    ///
+    /// 1. No tag could be found for `oid`
+    /// 2. The reference path for the `urn` could not be found (it defaults to
+    /// `rad/id` if not provided)
+    /// 3. The SHA of the tag was not the same as the resolved reference
+    /// 4. The `oid` was the [`zero`][`git2::Oid::zero`] SHA.
+    #[tracing::instrument(level = "debug", skip(self, urn), fields(urn = %urn), err)]
+    pub fn has_tag<Oid>(&self, urn: &Urn, oid: Oid) -> Result<bool, Error>
+    where
+        Oid: AsRef<git2::Oid> + Debug,
+    {
+        let oid = match self.has_object(oid)? {
+            Some(oid) => oid,
+            None => return Ok(false),
+        };
+
+        self.backend
+            .find_tag(oid)
+            .and_then(|tag| {
+                let tip = match self.get_urn_oid(urn)? {
+                    Some(tip) => tip,
+                    None => return Ok(false),
+                };
+                Ok(tip == tag.id())
+            })
+            .or_matches(is_not_found_err, || {
+                tracing::debug!("tag not found");
+                Ok(false)
+            })
+    }
+
+    fn get_urn_oid(&self, urn: &Urn) -> Result<Option<git2::Oid>, git2::Error> {
```

Comment:
Not sure... if you want to remove redundant code, I think `tip` does the same thing.